File: lib/engines/training_engine.dart
import 'dart:async';
import 'dart:math';

import '../models/training_config.dart';
import 'metronome_engine.dart';
import 'tone_generator.dart';

enum TrainingPhase { idle, countIn, running, rest, paused, finished }

class SignalEvent {
  final Direction direction;
  final DirectionSignalType signalType;

  SignalEvent(this.direction, this.signalType);
}

class TrainingState {
  final TrainingPhase phase;
  final int elapsedSeconds;
  final int totalSeconds;
  final int beatCount;
  final int signalCount;
  final int currentBpm;
  final int roundNumber;
  final int totalRounds;
  final Direction? lastDirection;

  TrainingState({
    this.phase = TrainingPhase.idle,
    this.elapsedSeconds = 0,
    this.totalSeconds = 0,
    this.beatCount = 0,
    this.signalCount = 0,
    this.currentBpm = 100,
    this.roundNumber = 1,
    this.totalRounds = 1,
    this.lastDirection,
  });
}

class TrainingEngine {
  final TrainingConfig config;
  final MetronomeEngine _metronome = MetronomeEngine();

  final _stateController = StreamController<TrainingState>.broadcast();
  final _signalController = StreamController<SignalEvent>.broadcast();
  final _beatController = StreamController<BeatEvent>.broadcast();
  final _tickController = StreamController<void>.broadcast();

  final Random _random = Random();
  Timer? _durationTimer;
  Timer? _bpmTimer;
  StreamSubscription<BeatEvent>? _beatSub;

  int _signalCount = 0;
  int _beatsSinceLastSignal = 0;
  int _elapsedSeconds = 0;
  int _roundNumber = 1;
  int _currentBpm;
  int _baseBpm;
  Direction? _lastDirection;
  TrainingPhase _phase = TrainingPhase.idle;

  TrainingEngine(this.config) : _currentBpm = config.bpm, _baseBpm = config.bpm;

  Stream<TrainingState> get stateStream => _stateController.stream;
  Stream<SignalEvent> get signalStream => _signalController.stream;
  Stream<BeatEvent> get beatStream => _beatController.stream;
  Stream<void> get tickStream => _tickController.stream;

  TrainingPhase get phase => _phase;
  int get signalCount => _signalCount;
  int get elapsedSeconds => _elapsedSeconds;
  int get currentBpm => _currentBpm;

  Future<void> start() async {
    _signalCount = 0;
    _beatsSinceLastSignal = 0;
    _elapsedSeconds = 0;
    _roundNumber = 1;
    _lastDirection = null;
    _currentBpm = config.bpm;
    _baseBpm = config.bpm;

    final intervalMs = config.beatIntervalMs;
    _metronome.configure(
      beatIntervalMs: intervalMs,
      beatsPerBar: config.beatsPerBar,
    );

    _phase = TrainingPhase.countIn;
    _emitState();

    _beatSub = _metronome.beatStream.listen(_onBeat);

    // 3-2-1 count-in
    for (int i = 3; i >= 1; i--) {
      _beatController.add(BeatEvent(
        beatNumber: i,
        isCountIn: true,
        isFirstBeat: i == 1,
      ));
      _tickController.add(null);
      await Future.delayed(Duration(milliseconds: intervalMs.round()));
    }

    _phase = TrainingPhase.running;
    _emitState();
    _metronome.start();
    _startDurationTracking();
    _startRandomBpmVariation();
  }

  void _startRandomBpmVariation() {
    if (!config.enableRandomBpm) return;
    // Vary BPM every 2-4 bars
    _bpmTimer = Timer.periodic(
      Duration(milliseconds: (config.beatIntervalMs * config.beatsPerBar * 2).round()),
      (_) => _varyBpm(),
    );
  }

  void _varyBpm() {
    final range = (config.bpm * config.randomBpmPercent / 100).round();
    final minBpm = (config.bpm - range).clamp(40, 500);
    final maxBpm = (config.bpm + range).clamp(40, 500);
    _currentBpm = minBpm + _random.nextInt(max(maxBpm - minBpm, 1));
    final intervalMs = 60000.0 / _currentBpm / config.beatSubdivision;
    _metronome.configure(beatIntervalMs: intervalMs, beatsPerBar: config.beatsPerBar);
    _emitState();
  }

  void _onBeat(BeatEvent beat) {
    _beatsSinceLastSignal++;

    _beatController.add(beat);
    _tickController.add(null);

    if (_beatsSinceLastSignal >= config.minBeatsToChange) {
      final maxBeats = config.maxBeatsToChange;
      final chance = _beatsSinceLastSignal >= maxBeats
          ? 1.0
          : 1.0 / (maxBeats - _beatsSinceLastSignal + 1);
      if (_random.nextDouble() < chance) {
        _fireSignal();
      }
    }
  }

  void _fireSignal() {
    _beatsSinceLastSignal = 0;
    _signalCount++;

    final directions = config.activeDirections;
    Direction dir;
    if (directions.length > 1 && _lastDirection != null) {
      final others = directions.where((d) => d != _lastDirection).toList();
      dir = others[_random.nextInt(others.length)];
    } else {
      dir = directions[_random.nextInt(directions.length)];
    }
    _lastDirection = dir;

    final signalType = _pickSignalType(dir);
    _signalController.add(SignalEvent(dir, signalType));
  }

  DirectionSignalType _pickSignalType(Direction dir) {
    if (config.signalSoundType == SignalSoundType.voiceDirection) {
      // Use direction-specific rapid two-note patterns
      switch (dir) {
        case Direction.forward: return DirectionSignalType.dirForward;
        case Direction.backward: return DirectionSignalType.dirBackward;
        case Direction.left: return DirectionSignalType.dirLeft;
        case Direction.right: return DirectionSignalType.dirRight;
      }
    }
    // Tone mode: single consistent alert sound for all directions
    return DirectionSignalType.alert;
  }

  void _startDurationTracking() {
    _durationTimer?.cancel();

    if (config.mode == TrainingMode.free) {
      _durationTimer = Timer.periodic(const Duration(seconds: 1), (_) {
        _elapsedSeconds++;
        _emitState();
      });
    } else if (config.mode == TrainingMode.timed) {
      final totalSecs = config.trainingDuration.inSeconds;
      _durationTimer = Timer.periodic(const Duration(seconds: 1), (_) {
        _elapsedSeconds++;
        _emitState();
        if (_elapsedSeconds >= totalSecs) stop();
      });
    } else if (config.mode == TrainingMode.progressive) {
      _durationTimer = Timer.periodic(const Duration(seconds: 1), (_) {
        _elapsedSeconds++;
        _emitState();
      });
      // Progressive overrides random BPM: increase every 20s
      Timer.periodic(const Duration(seconds: 20), (_) {
        if (_phase == TrainingPhase.running && _currentBpm < 500) {
          _baseBpm = min(500, _baseBpm + 5);
          _currentBpm = _baseBpm;
          final intervalMs = 60000.0 / _currentBpm / config.beatSubdivision;
          _metronome.configure(beatIntervalMs: intervalMs, beatsPerBar: config.beatsPerBar);
          _emitState();
        }
      });
    } else if (config.mode == TrainingMode.interval) {
      _startIntervalWorkPhase();
    }
  }

  void _startIntervalWorkPhase() {
    _phase = TrainingPhase.running;
    _emitState();
    _elapsedSeconds = 0;

    _durationTimer?.cancel();
    _durationTimer = Timer.periodic(const Duration(seconds: 1), (_) {
      _elapsedSeconds++;
      _emitState();
      if (_elapsedSeconds >= config.intervalWorkSeconds) {
        if (_roundNumber >= config.intervalRounds) {
          stop();
        } else {
          _startRestPhase();
        }
      }
    });
  }

  void _startRestPhase() {
    _phase = TrainingPhase.rest;
    _emitState();
    _elapsedSeconds = 0;

    _durationTimer?.cancel();
    _durationTimer = Timer.periodic(const Duration(seconds: 1), (_) {
      _elapsedSeconds++;
      _emitState();
      if (_elapsedSeconds >= config.intervalRestSeconds) {
        _roundNumber++;
        _startIntervalWorkPhase();
      }
    });
  }

  void stop() {
    _phase = TrainingPhase.finished;
    _emitState();
    _metronome.stop();
    _durationTimer?.cancel();
    _bpmTimer?.cancel();
    _beatSub?.cancel();
  }

  void dispose() {
    stop();
    _metronome.dispose();
    _stateController.close();
    _signalController.close();
    _beatController.close();
    _tickController.close();
  }

  void _emitState() {
    int totalSecs = 0;
    if (config.mode == TrainingMode.timed) {
      totalSecs = config.trainingDuration.inSeconds;
    }

    _stateController.add(TrainingState(
      phase: _phase,
      elapsedSeconds: _elapsedSeconds,
      totalSeconds: totalSecs,
      beatCount: _metronome.beatCount,
      signalCount: _signalCount,
      currentBpm: _currentBpm,
      roundNumber: _roundNumber,
      totalRounds: config.mode == TrainingMode.interval ? config.intervalRounds : 1,
      lastDirection: _lastDirection,
    ));
  }
}

File: wechat-mini-program/src/engines/trainingEngine.ts
import {
  TrainingConfig,
  TrainingMode,
  TrainingPhase,
  Direction,
  DirectionSignalType,
  SignalSoundType,
  getActiveDirections,
} from '../models/config';
import { Metronome, BeatEvent } from './metronome';

export interface TrainingState {
  phase: TrainingPhase;
  elapsedSeconds: number;
  totalSeconds: number;
  beatCount: number;
  signalCount: number;
  currentBpm: number;
  roundNumber: number;
  totalRounds: number;
  lastDirection: Direction | null;
}

export interface SignalEvent {
  direction: Direction;
  signalType: DirectionSignalType;
}

export interface CountdownEvent {
  remainingSeconds: number;
}

type EventHandler = {
  state?: (s: TrainingState) => void;
  signal?: (e: SignalEvent) => void;
  beat?: (e: BeatEvent) => void;
  tick?: () => void;
  countdown?: (e: CountdownEvent) => void;
  endBell?: () => void;
};

export class TrainingEngine {
  private config: TrainingConfig;
  private metronome: Metronome = new Metronome();
  private handler: EventHandler = {};

  private signalCount = 0;
  private beatsSinceLastSignal = 0;
  private elapsedSeconds = 0;
  private roundNumber = 1;
  private currentBpm: number;
  private baseBpm: number;
  private lastDirection: Direction | null = null;
  private phase: TrainingPhase = TrainingPhase.idle;
  private countdown10Fired = false;
  private countdown5Fired = false;

  private durationTimer: ReturnType<typeof setInterval> | null = null;
  private bpmTimer: ReturnType<typeof setInterval> | null = null;
  private boundOnBeat: (e: BeatEvent) => void;

  constructor(config: TrainingConfig) {
    this.config = config;
    this.currentBpm = config.bpm;
    this.baseBpm = config.bpm;
    this.boundOnBeat = this.onBeat.bind(this);
  }

  on(handler: EventHandler): void {
    this.handler = handler;
  }

  get isPaused(): boolean {
    return this.phase === TrainingPhase.paused;
  }

  get currentPhase(): TrainingPhase {
    return this.phase;
  }

  getCurrentBpm(): number {
    return this.currentBpm;
  }

  getElapsedSeconds(): number {
    return this.elapsedSeconds;
  }

  getSignalCount(): number {
    return this.signalCount;
  }

  async start(): Promise<void> {
    this.signalCount = 0;
    this.beatsSinceLastSignal = 0;
    this.elapsedSeconds = 0;
    this.roundNumber = 1;
    this.lastDirection = null;
    this.currentBpm = this.config.bpm;
    this.baseBpm = this.config.bpm;
    this.countdown10Fired = false;
    this.countdown5Fired = false;

    const intervalMs = this.beatIntervalMs();
    this.metronome.configure(intervalMs, this.config.beatsPerBar);
    this.metronome.onBeat(this.boundOnBeat);

    this.phase = TrainingPhase.countIn;
    this.emitState();

    // 3-2-1-开始 count-in
    for (let i = 3; i >= 0; i--) {
      const event: BeatEvent = {
        beatNumber: i === 0 ? 0 : i,
        isFirstBeat: i === 0,
        isCountIn: true,
      };
      this.handler.beat?.(event);
      this.handler.tick?.();
      await this.delay(intervalMs);
    }

    this.phase = TrainingPhase.running;
    this.emitState();
    this.metronome.start();
    this.startDurationTracking();
    this.startRandomBpmVariation();
  }

  pause(): void {
    if (this.phase !== TrainingPhase.running && this.phase !== TrainingPhase.rest) return;
    this.phase = TrainingPhase.paused;
    this.metronome.stop();
    this.clearTimers();
    this.emitState();
  }

  resume(): void {
    if (this.phase !== TrainingPhase.paused) return;
    this.phase = TrainingPhase.running;
    this.metronome.start();
    this.startDurationTracking();
    this.startRandomBpmVariation();
    this.emitState();
  }

  stop(): void {
    this.phase = TrainingPhase.finished;
    this.emitState();
    this.metronome.stop();
    this.metronome.removeListener(this.boundOnBeat);
    this.clearTimers();
  }

  dispose(): void {
    this.stop();
    this.metronome.removeListener(this.boundOnBeat);
  }

  private clearTimers(): void {
    if (this.durationTimer !== null) { clearInterval(this.durationTimer); this.durationTimer = null; }
    if (this.bpmTimer !== null) { clearInterval(this.bpmTimer); this.bpmTimer = null; }
  }

  private delay(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  private beatIntervalMs(): number {
    return 60000 / this.currentBpm / this.config.beatSubdivision;
  }

  private onBeat(beat: BeatEvent): void {
    this.beatsSinceLastSignal++;
    this.handler.beat?.(beat);
    this.handler.tick?.();

    if (this.beatsSinceLastSignal >= this.config.minBeatsToChange) {
      const maxBeats = this.config.maxBeatsToChange;
      const chance = this.beatsSinceLastSignal >= maxBeats
        ? 1
        : 1 / (maxBeats - this.beatsSinceLastSignal + 1);
      if (Math.random() < chance) {
        this.fireSignal();
      }
    }
  }

  private fireSignal(): void {
    this.beatsSinceLastSignal = 0;
    this.signalCount++;

    const directions = getActiveDirections(this.config.directionCount);
    let dir: Direction;
    if (directions.length > 1 && this.lastDirection !== null) {
      const others = directions.filter(d => d !== this.lastDirection);
      dir = others[Math.floor(Math.random() * others.length)];
    } else {
      dir = directions[Math.floor(Math.random() * directions.length)];
    }
    this.lastDirection = dir;

    const signalType = this.pickSignalType(dir);
    this.handler.signal?.({ direction: dir, signalType });
  }

  private pickSignalType(dir: Direction): DirectionSignalType {
    if (this.config.signalSoundType === SignalSoundType.voiceDirection) {
      switch (dir) {
        case Direction.forward: return DirectionSignalType.dirForward;
        case Direction.backward: return DirectionSignalType.dirBackward;
        case Direction.left: return DirectionSignalType.dirLeft;
        case Direction.right: return DirectionSignalType.dirRight;
      }
    }
    return DirectionSignalType.alert;
  }

  private startRandomBpmVariation(): void {
    if (!this.config.enableRandomBpm) return;
    this.bpmTimer = setInterval(() => {
      if (this.phase !== TrainingPhase.running) return;
      const range = Math.round(this.config.bpm * this.config.randomBpmPercent / 100);
      const minBpm = Math.max(40, this.config.bpm - range);
      const maxBpm = Math.min(500, this.config.bpm + range);
      this.currentBpm = minBpm + Math.floor(Math.random() * Math.max(maxBpm - minBpm, 1));
      this.metronome.configure(this.beatIntervalMs(), this.config.beatsPerBar);
      this.emitState();
    }, this.beatIntervalMs() * this.config.beatsPerBar * 2);
  }

  private startDurationTracking(): void {
    this.clearTimers();

    if (this.config.mode === TrainingMode.free) {
      this.durationTimer = setInterval(() => {
        this.elapsedSeconds++;
        this.emitState();
      }, 1000);
    } else if (this.config.mode === TrainingMode.timed) {
      const totalSecs = this.config.trainingDuration * 60;
      this.durationTimer = setInterval(() => {
        this.elapsedSeconds++;
        this.emitState();
        this.checkCountdown(totalSecs);
        if (this.elapsedSeconds >= totalSecs) {
          this.handler.endBell?.();
          this.stop();
        }
      }, 1000);
    } else if (this.config.mode === TrainingMode.progressive) {
      this.durationTimer = setInterval(() => {
        this.elapsedSeconds++;
        this.emitState();
      }, 1000);
      setInterval(() => {
        if (this.phase === TrainingPhase.running && this.currentBpm < 500) {
          this.baseBpm = Math.min(500, this.baseBpm + 5);
          this.currentBpm = this.baseBpm;
          this.metronome.configure(this.beatIntervalMs(), this.config.beatsPerBar);
          this.emitState();
        }
      }, 20000);
    } else if (this.config.mode === TrainingMode.interval) {
      this.startIntervalWorkPhase();
    }
  }

  private checkCountdown(totalSecs: number): void {
    const remaining = totalSecs - this.elapsedSeconds;
    if (remaining <= 10 && remaining > 5 && !this.countdown10Fired) {
      this.countdown10Fired = true;
      this.handler.countdown?.({ remainingSeconds: remaining });
    } else if (remaining <= 5 && remaining > 0 && !this.countdown5Fired) {
      this.countdown5Fired = true;
      this.handler.countdown?.({ remainingSeconds: remaining });
    }
  }

  private startIntervalWorkPhase(): void {
    this.phase = TrainingPhase.running;
    this.emitState();
    this.elapsedSeconds = 0;
    this.countdown10Fired = false;
    this.countdown5Fired = false;

    this.durationTimer = setInterval(() => {
      this.elapsedSeconds++;
      this.emitState();
      this.checkCountdown(this.config.intervalWorkSeconds);
      if (this.elapsedSeconds >= this.config.intervalWorkSeconds) {
        if (this.roundNumber >= this.config.intervalRounds) {
          this.handler.endBell?.();
          this.stop();
        } else {
          this.startRestPhase();
        }
      }
    }, 1000);
  }

  private startRestPhase(): void {
    this.phase = TrainingPhase.rest;
    this.emitState();
    this.elapsedSeconds = 0;
    this.countdown10Fired = false;
    this.countdown5Fired = false;

    this.durationTimer = setInterval(() => {
      this.elapsedSeconds++;
      this.emitState();
      if (this.elapsedSeconds >= this.config.intervalRestSeconds) {
        this.roundNumber++;
        this.startIntervalWorkPhase();
      }
    }, 1000);
  }

  private emitState(): void {
    let totalSecs = 0;
    if (this.config.mode === TrainingMode.timed) {
      totalSecs = this.config.trainingDuration * 60;
    }
    this.handler.state?.({
      phase: this.phase,
      elapsedSeconds: this.elapsedSeconds,
      totalSeconds: totalSecs,
      beatCount: 0,
      signalCount: this.signalCount,
      currentBpm: this.currentBpm,
      roundNumber: this.roundNumber,
      totalRounds: this.config.mode === TrainingMode.interval ? this.config.intervalRounds : 1,
      lastDirection: this.lastDirection,
    });
  }
}
